include/s3select_functions.h

```diff
@@ -643,226 +774,415 @@ struct _fn_to_timestamp : public base_function
       throw base_s3select_exception("to_timestamp first argument must be string");  //can skip current row
     }
 
-    bsc::parse_info<> info_dig = bsc::parse(v_str.str(), d_yyyymmdd_dig >> *(separator) >> d_time_dig);
+    bsc::parse_info<> info_dig = bsc::parse(v_str.str(), d_date_time);
 
     if(datetime_validation()==false or !info_dig.full)
     {
       throw base_s3select_exception("input date-time is illegal");
     }
 
-    new_ptime = boost::posix_time::ptime(boost::gregorian::date(yr, mo, dy),
-                                         boost::posix_time::hours(hr) + boost::posix_time::minutes(mn) + boost::posix_time::seconds(sc));
-
-    result->set_value(&new_ptime);
+    #if NANO_SEC
+      //TODO: Include timezone hours(tz_hr) and timezone minutes(tz_mn) in date time calculation.
+      new_ptime = boost::posix_time::ptime(boost::gregorian::date(yr, mo, dy),
+                          boost::posix_time::hours(hr) +
+                          boost::posix_time::minutes(mn) +
+                          boost::posix_time::seconds(sc) +
+                          boost::posix_time::nanoseconds(frac_sec*1000000000));
+
+      result->set_value(&new_ptime);
+    #else
+      if (frac_sec >= 0 && frac_sec < 10)
+              frac_sec = frac_sec * 100000;
+      else if (frac_sec >= 10 && frac_sec < 100)
+              frac_sec = frac_sec * 10000;
+      else if (frac_sec >= 100 && frac_sec < 1000)
+              frac_sec = frac_sec * 1000;
+      else if (frac_sec >= 1000 && frac_sec < 10000)
+              frac_sec = frac_sec * 100;
+      else if (frac_sec >= 0 && frac_sec < 100000)
+              frac_sec = frac_sec * 10;
+
+      //TODO: Include timezone hours(tz_hr) and timezone minutes(tz_mn) in date time calculation.
+      new_ptime = boost::posix_time::ptime(boost::gregorian::date(yr, mo, dy),
+                          boost::posix_time::time_duration(hr, mn, sc, frac_sec));
+
+      result->set_value(&new_ptime);
+    #endif
 
     return true;
   }
 
 };
 
-struct _fn_extact_from_timestamp : public base_function
+void extract_timestamp_param_validation(bs_stmt_vec_t*& args, base_statement*& ts)
 {
+  auto iter = args->begin();
+    int args_size = args->size();
+
+    if (args_size < 1)
+    {
+      throw base_s3select_exception("to_timestamp should have 2 parameters");
+    }
 
+    ts = *iter;
+
+    if(ts->eval().is_timestamp()== false)
+    {
+      throw base_s3select_exception("second parameter is not timestamp");
+    }
+}
+
+struct _fn_extract_year_from_timestamp : public base_function
+{
   boost::posix_time::ptime new_ptime;
 
-  value val_date_part;
+  bool operator()(bs_stmt_vec_t* args, variable* result) override
+  {
+    base_statement* ts;
+    extract_timestamp_param_validation(args, ts);
```

Comment:
/extract_timestamp_param_validation/ should reside in base-object, doing so will enable reuse value.

the /ts/ arg is actually called twice. per each row/column.